modules/dnn/misc/python/test/test_dnn.py

```diff
@@ -226,6 +226,73 @@ def test_async(self):
                 self.assertTrue(ret)
                 normAssert(self, refs[i], result, 'Index: %d' % i, 1e-10)
 
+    def test_custom_layer(self):
+        class CropLayer(object):
+            def __init__(self, params, blobs):
+                self.xstart = 0
+                self.xend = 0
+                self.ystart = 0
+                self.yend = 0
+            # Our layer receives two inputs. We need to crop the first input blob
+            # to match a shape of the second one (keeping batch size and number of channels)
+            def getMemoryShapes(self, inputs):
+                inputShape, targetShape = inputs[0], inputs[1]
+                batchSize, numChannels = inputShape[0], inputShape[1]
+                height, width = targetShape[2], targetShape[3]
+                self.ystart = (inputShape[2] - targetShape[2]) // 2
+                self.xstart = (inputShape[3] - targetShape[3]) // 2
+                self.yend = self.ystart + height
+                self.xend = self.xstart + width
+                return [[batchSize, numChannels, height, width]]
+            def forward(self, inputs):
+                return [inputs[0][:,:,self.ystart:self.yend,self.xstart:self.xend]]
+
+        cv.dnn_registerLayer('CropCaffe', CropLayer)
+        proto = 'name: "TestCrop"\n' \
+        'input: "input"\n' \
+        'input_shape\n' \
+        '{\n' \
+        '    dim: 1\n' \
+        '    dim: 2\n' \
+        '    dim: 5\n' \
+        '    dim: 5\n' \
+        '}\n' \
+        'input: "roi"\n' \
+        'input_shape\n' \
+        '{\n' \
+        '    dim: 1\n' \
+        '    dim: 2\n' \
+        '    dim: 3\n' \
+        '    dim: 3\n' \
+        '}\n' \
+        'layer {\n' \
+        '  name: "Crop"\n' \
+        '  type: "CropCaffe"\n' \
+        '  bottom: "input"\n' \
+        '  bottom: "roi"\n' \
+        '  top: "Crop"\n' \
+        '} '
+
+        net = cv.dnn.readNetFromCaffe(str.encode(proto))
```

Comment:
There are several functions with the same name. Code should call the right overload.
Try to wrap input into `bytearray(proto.encode())` (just string is a file name).

P.S. Revert multiline string back.